build: fix api guardian path for windows

alan-agius4 opened the pull request:
At the moment, `path.posix.relative` will break paths in windows as it will return something like
```
Error: Source file "../C:/users/alag/_bazel_alag/3tbqurya/execroot/angular/bazel-out/x64_windows-fastbuild/bin/packages/core/core.d.ts" not found
```

Remove `\r`  followed by `\n` as in windows these causes golden files to be different.

DevVersion commented on tools/ts-api-guardian/lib/cli.ts:
```diff
@@ -191,7 +191,7 @@ function resolveBazelFilePath(fileName: string): string {
   // runfile, we need to use `require.resolve` which handles runfiles properly on Windows.
   if (process.env['BAZEL_TARGET']) {
-    return path.posix.relative(process.cwd(), require.resolve(fileName));
+    return require.resolve(fileName);
```
If we don't resolve the path relatively, it causes the `verify.patch` to be thrown off. 

https://github.com/angular/angular/blob/b184b116c1c3a56bd7f2769768e32c54b413bd59/tools/ts-api-guardian/test/fixtures/verify.patch.

That was one of the reasons I switched to a relative path, besides keeping the patch output non-breaking.

gkalpak replied:
Maybe use [canonical-path][1]? :wink:

[1]: https://www.npmjs.com/package/canonical-path

alan-agius4 replied:
I will try to see it `path.relative` works for both linux and windows. In my case it did work.

alan-agius4 replied:
Why am I having lint issues on files that I pull from master?

alan-agius4 committed:
build: fix api guardian path for windows

At the moment, `path.posix.relative` will break paths in windows as it will return something like
```
Error: Source file "../C:/users/alag/_bazel_alag/3tbqurya/execroot/angular/bazel-out/x64_windows-fastbuild/bin/packages/core/core.d.ts" not found
```

alexeagle commented on tools/ts-api-guardian/lib/serializer.ts:
```diff
@@ -381,7 +381,12 @@ const memberDeclarationOrder: {[key: number]: number} = {
 
 function stripEmptyLines(text: string): string {
-  return text.split('\n').filter(x => !!x.length).join('\n');
+  return text.split('\n')
```
simplify?
```
> "a.b,c".split(/[.,]/)
[ 'a', 'b', 'c' ]
```

alan-agius4 replied:
This should work!, will try it later today/tmr.

alan-agius4 replied:
Updated

gkalpak commented on tools/ts-api-guardian/lib/serializer.ts:
```diff
@@ -381,7 +381,7 @@ const memberDeclarationOrder: {[key: number]: number} = {
 
 function stripEmptyLines(text: string): string {
-  return text.split('\n').filter(x => !!x.length).join('\n');
+  return text.split(/[\n\r]/).filter(x => !!x.length).join('\n');
```
This works, but only because you filter out empty "lines". If you were doing some other kind of processing (e.g. counting lines or transforming lines) you would get incorrect results.

A more common way to split on both `\n` and `\r\n` is: `/\r?\n/`

alan-agius4 replied:
@gkalpak, Indeed, Thanks.  Updated the `RegExp` as suggested.

alan-agius4 committed:
build: avoid writing `\r` in windows

In windows when using readFile `\n` are replaced with `\r\n` which causes issues when comparing golden files

alan-agius4 commented:
CI is flaky
